Add CUDA-only mode job to the CI

SimeonEhrig opened the pull request:
related to this problem: https://github.com/alpaka-group/alpaka/pull/2479#issue-2857604635

fwyzard commented:
Is the error

> ```
> /builds/hzdr/crp/alpaka/example/useBLASInAlpaka/useCuBLASInAlpaka/src/useCuBLASInAlpaka.cpp:11:10: fatal error: cublas_v2.h: No such file or directory
>    11 | #include <cublas_v2.h>
>       |          ^~~~~~~~~~~~~
> ```

actually related to this PR ?

SimeonEhrig commented:
> Is the error
> 
> > ```
> > /builds/hzdr/crp/alpaka/example/useBLASInAlpaka/useCuBLASInAlpaka/src/useCuBLASInAlpaka.cpp:11:10: fatal error: cublas_v2.h: No such file or directory
> >    11 | #include <cublas_v2.h>
> >       |          ^~~~~~~~~~~~~
> > ```
> 
> actually related to this PR ?

No, the error was introduced in another PR. Looks like the CMake is wrong. But without CI job, would could not detect it. I will fix it later.

SimeonEhrig commented:
> > Is the error
> > > ```
> > > /builds/hzdr/crp/alpaka/example/useBLASInAlpaka/useCuBLASInAlpaka/src/useCuBLASInAlpaka.cpp:11:10: fatal error: cublas_v2.h: No such file or directory
> > >    11 | #include <cublas_v2.h>
> > >       |          ^~~~~~~~~~~~~
> > > ```
> > 
> > 
> > actually related to this PR ?
> 
> No, the error was introduced in another PR. Looks like the CMake is wrong. But without CI job, would could not detect it. I will fix it later.

I was wrong. The problem is the container. cuBlas is not installed. Now I have to check why. I thought it is part of the basic installation of the CUDA SDK.

fwyzard commented:
> I was wrong. The problem is the container. cuBlas is not installed. Now I have to check why. I thought it is part of the basic installation of the CUDA SDK.

But, how do the other CUDA CI tests work, then ?

SimeonEhrig commented:
> > I was wrong. The problem is the container. cuBlas is not installed. Now I have to check why. I thought it is part of the basic installation of the CUDA SDK.
> 
> But, how do the other CUDA CI tests work, then ?

Other tests does not require the cuBlas library and until yet we didn't test this example, because we had no cuda only test.

fwyzard commented:
> Other tests does not require the cuBlas library

Ah !

SimeonEhrig committed:
add cuda only mode job to the CI

CI_FILTER: linux_nvcc12.5_gcc13_release_cuda_only_mode

fwyzard commented:
Same error as https://github.com/alpaka-group/alpaka/pull/2503#issuecomment-2967341691 :-/

SimeonEhrig committed:
fix compiler errors for cuda only build

- two tests are still broken: matrixMulMdSpan and mathTest

CI_FILTER: linux_nvcc12.5_gcc13_release_cuda_only_mode

SimeonEhrig committed:
run tests in cuda only CI job

CI_FILTER: linux_nvcc12.5_gcc13_release_cuda_only_mode

SimeonEhrig commented:
@chillenzer Looks like that the bug from PR #2479 still exist: https://gitlab.com/hzdr/crp/alpaka/-/jobs/10341320002
Not we get a better error message.

That is a cmake config, which you can use to reproduce it locally.

```bash
cmake -DCMAKE_BUILD_TYPE=RELEASE -DCMAKE_CXX_COMPILER=g++ -Dalpaka_ACC_CPU_B_SEQ_T_SEQ_ENABLE=OFF -Dalpaka_ACC_GPU_CUDA_ONLY_MODE=ON -Dalpaka_ENABLE_WERROR=ON -Dalpaka_ACC_GPU_CUDA_ENABLE=ON -Dalpaka_BUILD_EXAMPLES=ON -DBUILD_TESTING=ON -Dalpaka_USE_MDSPAN=FETCH -Dalpaka_BUILD_BENCHMARKS=ON -Dalpaka_CHECK_HEADERS=ON
```

mehmetyusufoglu commented on example/bufferCopy/src/bufferCopy.cpp:
```diff
@@ -9,6 +9,8 @@
 #include <cstdint>
 #include <iostream>
 
+#ifdef ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLED
```
Only expected to run for this backend ?

SimeonEhrig replied:
No, if I don't define it wrong the serial backend needs to be enabled and all other backends can be.

mehmetyusufoglu replied:
ok, that means it is not compiled for cuda-only.

SimeonEhrig replied:
That's correct. The examples requires an CPU accelerator for the host, to enqueue tasks. Therefore they cannot work in CUDA only mode, where all accelerators except the CUDA is disabled.

SimeonEhrig committed:
change data type of matrix multiplication example to avoid complex handling of summation error of float

- fixed break statement in double loop, which was not working like expected

fwyzard commented:
Why the change in data type ?

chillenzer commented:
> @chillenzer Looks like that the bug from PR #2479 still exist: https://gitlab.com/hzdr/crp/alpaka/-/jobs/10341320002 Not we get a better error message.
> 
> That is a cmake config, which you can use to reproduce it locally.
> 
> ```shell
> cmake -DCMAKE_BUILD_TYPE=RELEASE -DCMAKE_CXX_COMPILER=g++ -Dalpaka_ACC_CPU_B_SEQ_T_SEQ_ENABLE=OFF -Dalpaka_ACC_GPU_CUDA_ONLY_MODE=ON -Dalpaka_ENABLE_WERROR=ON -Dalpaka_ACC_GPU_CUDA_ENABLE=ON -Dalpaka_BUILD_EXAMPLES=ON -DBUILD_TESTING=ON -Dalpaka_USE_MDSPAN=FETCH -Dalpaka_BUILD_BENCHMARKS=ON -Dalpaka_CHECK_HEADERS=ON
> ```

Ah, I knew that there was something. If this is urgent, someone else should pick it up please. This would land on a very full to-do list currently and is not highest priority.

SimeonEhrig commented:
> Why the change in data type ?

I get summation errors if I compare serial implementation and a parallelized implementation because of the low precession of `float`. The delta is about 10e7. I could spend some time to solve the problem or simply change the data type which does not have the problem. I think for the example it is fine, because it should demonstrate memory access with a well know algorithm.

mehmetyusufoglu commented:
Changing ALPAKA_FN_HOST_ACC to ALPAKA_FN_ACC at LambdaMathTestTemplate seems solved the problem. I will create a PR

mehmetyusufoglu commented:
> > @chillenzer Looks like that the bug from PR #2479 still exist: https://gitlab.com/hzdr/crp/alpaka/-/jobs/10341320002 Not we get a better error message.
> > That is a cmake config, which you can use to reproduce it locally.
> > ```shell
> > cmake -DCMAKE_BUILD_TYPE=RELEASE -DCMAKE_CXX_COMPILER=g++ -Dalpaka_ACC_CPU_B_SEQ_T_SEQ_ENABLE=OFF -Dalpaka_ACC_GPU_CUDA_ONLY_MODE=ON -Dalpaka_ENABLE_WERROR=ON -Dalpaka_ACC_GPU_CUDA_ENABLE=ON -Dalpaka_BUILD_EXAMPLES=ON -DBUILD_TESTING=ON -Dalpaka_USE_MDSPAN=FETCH -Dalpaka_BUILD_BENCHMARKS=ON -Dalpaka_CHECK_HEADERS=ON
> > ```
> 
> Ah, I knew that there was something. If this is urgent, someone else should pick it up please. This would land on a very full to-do list currently and is not highest priority.

I am checking.

SimeonEhrig committed:
fix mathTest for cuda-only build

thanks to @mehmetyusufoglu for the solution

SimeonEhrig commented:
> Changing ALPAKA_FN_HOST_ACC to ALPAKA_FN_ACC at LambdaMathTestTemplate seems solved the problem. I will create a PR

I tested it locally. Looks good. Thanks for the fix.

SimeonEhrig committed:
disable all code which uses `std::atomic_ref` or `boost::atomic_ref` if no CPU backend is enabled

During testing the CUDA only build, I realized that the atomic_ref code was compiled all the time. The code is only used for the CPU backends. Because the check if `std::atomic_ref` is available is only enabled, if a CPU backend is enabled, the GPU backends uses the fallback `boost::atomic_ref` all the time. Therefore there was a hidden dependency on boost.

SimeonEhrig committed:
disable the zeroDimBuffer test for CUDA and HIP only mode

- For a successful test, we need to enable the serial CPU backend. The serial CPU backend requires  `std::atomic_ref` or `boost::atomic_ref`. Both introduces a dependency which is not required for the production code. It is only required for testing. Therefore disable the test of the CUDA and HIP only mode.

fwyzard commented on example/matrixMulWithMdspan/src/matrixMulMdSpan.cpp:
```diff
@@ -165,17 +165,19 @@ auto example(TAccTag const&) -> int
+            for(Idx j = 0; j < N; ++j)
             {
-                success = false;
-                break;
```
why remove the `break` ?

SimeonEhrig replied:
I thought it is not working, because we are in double loop and we exit only the inner loop. But this was a mistake on my side. We are only in a single loop. The new implementation does the same like the old one. Maybe a little bit more inefficient.

fwyzard commented on test/integ/cudaOnly/src/cudaNativeFunctions.cpp:
```diff
@@ -14,7 +14,7 @@
 #        pragma clang diagnostic ignored "-Wmissing-prototypes"
 #    endif
-__device__ auto userDefinedThreadFence() -> void
+__device__ static auto userDefinedThreadFence() -> void
```
just curious - why `static` ?

SimeonEhrig replied:
There was the warning, that `userDefinedThreadFence` needs to be defined. I think it means a definition in a header. I used static to change visibility to hidden.

fwyzard replied:
Mhm, that's what the `pragma clang diagnostic` right above were supposed to silence.

SimeonEhrig replied:
GCC throw the exception
